borg/archiver.py

```diff
@@ -353,8 +353,20 @@ def do_extract(self, args):
         sparse = args.sparse
         strip_components = args.strip_components
         dirs = []
-        for item in archive.iter_items(lambda item: matcher.match(item[b'path']), preload=True):
+        partial_extract = not matcher.empty() or strip_components
+        hardlink_masters = {} if partial_extract else None
+
+        def item_is_hardlink_master(item):
+            return (partial_extract and stat.S_ISREG(item[b'mode']) and
```

Comment:
nice name, but not totally true. if partial_extract is False, it returns False even if it is a hardlink_master.